File: cross-platform/src/core/storage.ts
/**
 * What the application is costing on disk, ported from `Storage.swift`.
 *
 * **Almost all of it is audio, and almost all of the audio is derivable** --
 * ninety-seven percent of the footprint is output the app can make again from
 * inputs it already has. So the honest offer is not "free up space", which
 * invites the listener to gamble, but a statement of what each pile *costs to
 * lose*.
 *
 * **Nothing here ever deletes a directory, and nothing ever deletes writing.**
 * A render directory carries its own `notes.md`, so purging one wholesale would
 * take the listener's notes with it. Purging removes files it has named, all of
 * them audio, and leaves the folder, the manifest and every word standing.
 */
import { existsSync, readdirSync, readFileSync, statSync, unlinkSync } from "fs";
import { join, basename, extname } from "path";
import { isCurrent, items } from "./renderPlan.js";
import { decodeManifest } from "./sessionManifest.js";
import type { Library } from "./library.js";

export type StorageKind =
  | "supersededTakes" | "currentTakes" | "assembledWithRecipe"
  | "assembledWithoutRecipe" | "assembledRetiredVoice" | "recycleBin" | "voicePreviews";

/** The order the report presents them in, which is `CaseIterable`'s order. */
export const storageKinds: StorageKind[] = [
  "supersededTakes", "currentTakes", "assembledWithRecipe",
  "assembledWithoutRecipe", "assembledRetiredVoice", "recycleBin", "voicePreviews",
];

export const storageTitle = (k: StorageKind): string => ({
  supersededTakes: "Superseded narration",
  currentTakes: "Current narration",
  assembledWithRecipe: "Assembled tapes",
  assembledWithoutRecipe: "Assembled tapes without a recipe",
  assembledRetiredVoice: "Assembled tapes in a retired voice",
  recycleBin: "Recently deleted",
  voicePreviews: "Voice previews",
}[k]);

/** What deleting this actually costs. Never "nothing" unless it is nothing. */
export const storageConsequence = (k: StorageKind): string => ({
  supersededTakes: "Already out of date. The queue re-renders these anyway, so this costs nothing.",
  currentTakes: "Still valid. Deleting them means rendering them again before the next assembly.",
  assembledWithRecipe: "Rebuildable exactly from the recipe that made them.",
  assembledWithoutRecipe: "These predate frozen recipes. They can be built again from their plan, but not identically.",
  assembledRetiredVoice: "Spoken by a voice that is no longer installed. They still play, but nothing can make them again -- rebuilding produces a different voice.",
  recycleBin: "Already deleted, kept for thirty days. Emptying now is permanent.",
  voicePreviews: "A minute to render again.",
}[k]);

/** True when nothing at all is lost -- not merely nothing important. */
export const costsNothing = (k: StorageKind): boolean => k === "supersededTakes";

export interface StorageGroup { kind: StorageKind; files: string[]; bytes: number }
export interface StorageReport { groups: StorageGroup[]; totalBytes: number }

export const reclaimableBytes = (r: StorageReport): number =>
  r.groups.reduce((n, g) => n + g.bytes, 0);

export const storageGroup = (r: StorageReport, k: StorageKind): StorageGroup | undefined =>
  r.groups.find(g => g.kind === k);

const sizeOf = (p: string): number => {
  try { return statSync(p).size; } catch { return 0; }
};
const listDir = (d: string): string[] => {
  try { return readdirSync(d); } catch { return []; }
};
const isDir = (p: string): boolean => {
  try { return statSync(p).isDirectory(); } catch { return false; }
};
const readText = (p: string): string | undefined => {
  try { return readFileSync(p, "utf8"); } catch { return undefined; }
};

/** Sources the render queue can rebuild a take from, keyed by output name. */
function sourcesByOutput(library: Library): Map<string, string> {
  const out = new Map<string, string>();
  // Every authored file, at every verbosity it was written at: a segment with a
  // v1 and a v3 body produces two differently named takes, and missing one
  // would report a live take as orphaned.
  const files: string[] = library.focus.flatMap(f => f.scripts);
  for (const ref of [...library.segments, ...library.continuousSegments]) {
    files.push(ref.path);
    files.push(...Object.values(ref.verbosityFiles));
  }
  for (const file of files) {
    const source = readText(file);
    if (source === undefined) continue;
    for (const item of items(file, source)) out.set(item.outputName, source);
  }
  return out;
}

/** `<session-id>-announcement.takeN.wav` -> `<session-id>`. */
export function announcementSession(name: string): string | undefined {
  const i = name.indexOf("-announcement.take");
  return i < 0 ? undefined : name.slice(0, i);
}

export function directorySize(dir: string): number {
  let total = 0;
  const walk = (d: string): void => {
    for (const name of listDir(d)) {
      const p = join(d, name);
      if (isDir(p)) walk(p); else total += sizeOf(p);
    }
  };
  walk(dir);
  return total;
}

/** Measure. Reads the disk and decides nothing on the listener's behalf. */
export function measure(o: {
  root: string; library: Library; renderKey: string; voice: string;
}): StorageReport {
  const { root, library, renderKey, voice } = o;
  const groups = new Map<StorageKind, { files: string[]; bytes: number }>();
  const add = (kind: StorageKind, url: string): void => {
    const g = groups.get(kind) ?? { files: [], bytes: 0 };
    g.files.push(url); g.bytes += sizeOf(url);
    groups.set(kind, g);
  };

  // Narration takes, split by whether they are still worth anything.
  //
  // **Every voice's directory, not just the selected one.** Retiring a voice
  // leaves its renders behind -- a gigabyte of them, in the case that prompted
  // this -- and measuring only the current voice made exactly the pile this
  // panel exists to find invisible.
  const sources = sourcesByOutput(library);
  const liveRenders = new Set(library.focus.flatMap(f => f.renders).map(p => basename(p)));
  const known = new Set(library.voices.map(v => v.name));
  const renderedRoot = join(root, "segments-rendered");
  for (const owner of listDir(renderedRoot).sort()) {
    const takeDir = join(renderedRoot, owner);
    if (!isDir(takeDir)) continue;
    const retired = !known.has(owner);
    for (const name of listDir(takeDir).sort()) {
      const url = join(takeDir, name);
      if (extname(url) !== ".wav") continue;
      if (retired) { add("supersededTakes", url); continue; }
      const source = sources.get(name);
      if (source !== undefined) {
        const current = owner === voice && isCurrent(name, source, takeDir, renderKey);
        add(current ? "currentTakes" : "supersededTakes", url);
      } else {
        const session = announcementSession(name);
        // An announcement belongs to one assembled tape and is current for as
        // long as that tape is.
        if (session !== undefined && liveRenders.has(session)) add("currentTakes", url);
        else add("supersededTakes", url);
      }
    }
  }

  // Assembled tapes. The audio only -- never the folder, never the notes.
  const installedVoices = new Set(library.voices.map(v => v.name));
  const recipes = new Set(listDir(join(root, "memory/sessions"))
    .filter(n => extname(n) === ".json").map(n => basename(n, ".json")));
  for (const dir of library.focus.flatMap(f => f.renders)) {
    const wav = join(dir, "session.wav");
    if (!existsSync(wav)) continue;
    // **Which voice spoke it outranks whether a recipe exists.** A recipe
    // promises an exact rebuild, and that promise is void the moment the voice
    // it names is gone: the words come back, the voice does not.
    const manifestText = readText(join(dir, "manifest.json"));
    let spokenBy: string | undefined;
    if (manifestText !== undefined) {
      try { spokenBy = decodeManifest(JSON.parse(manifestText)).voice; } catch { spokenBy = undefined; }
    }
    if (spokenBy !== undefined && spokenBy !== "" && !installedVoices.has(spokenBy)) {
      add("assembledRetiredVoice", wav);
    } else {
      add(recipes.has(basename(dir)) ? "assembledWithRecipe" : "assembledWithoutRecipe", wav);
    }
  }

  // The thirty-day bin, and previews.
  const walkBin = (d: string): void => {
    for (const name of listDir(d).sort()) {
      const p = join(d, name);
      if (isDir(p)) walkBin(p);
      else if (name !== "index.json") add("recycleBin", p);
    }
  };
  walkBin(join(root, "memory/deleted"));
  for (const v of library.voices) {
    const preview = join(v.path, "preview.wav");
    if (existsSync(preview)) add("voicePreviews", preview);
  }

  const ordered: StorageGroup[] = [];
  for (const kind of storageKinds) {
    const g = groups.get(kind);
    if (g && g.files.length > 0) ordered.push({ kind, files: g.files, bytes: g.bytes });
  }
  return { groups: ordered, totalBytes: directorySize(root) };
}

/**
 * Delete the named files. Returns the bytes actually freed.
 *
 * A directory here would mean the report had strayed, so it is refused rather
 * than recursed into -- the one place this could take a folder, and with it a
 * listener's notes.
 */
export function purge(report: StorageReport, kinds: Set<StorageKind>): number {
  let freed = 0;
  for (const group of report.groups) {
    if (!kinds.has(group.kind)) continue;
    for (const url of group.files) {
      if (isDir(url)) continue;
      const bytes = sizeOf(url);
      try { unlinkSync(url); freed += bytes; } catch { /* already gone */ }
    }
  }
  return freed;
}

File: Sources/GatewayCore/Storage.swift
import Foundation

/// What the application is costing on disk, and what it may delete to weigh
/// less.
///
/// **Almost all of it is audio, and almost all of the audio is derivable.**
/// On the owner's own machine the library of authored words is 91 MB and the
/// record of their practice is 156 KB, against 3.2 GB of rendered narration
/// and assembled tapes -- ninety-seven percent of the footprint is output that
/// the app can make again from inputs it already has.
///
/// So the honest offer is not "free up space", which invites the listener to
/// gamble, but a statement of what each pile *costs to lose*. Superseded takes
/// cost nothing: they are already invalid and the queue will re-render them
/// regardless. A current take costs the minutes to render it again. An
/// assembled tape with a frozen recipe costs the minutes to rebuild it exactly;
/// one without a recipe cannot be rebuilt identically, and says so.
///
/// **Nothing here ever deletes a directory, and nothing ever deletes writing.**
/// A render directory carries its own `notes.md`, so purging one wholesale
/// would take the listener's notes about the tape with it. Purging removes
/// files it has named, all of them audio, and leaves the folder, the manifest
/// and every word standing.
public enum StorageKind: String, CaseIterable, Sendable, Codable {
    case supersededTakes
    case currentTakes
    case assembledWithRecipe
    case assembledWithoutRecipe
    case assembledRetiredVoice
    case recycleBin
    case voicePreviews

    public var title: String {
        switch self {
        case .supersededTakes: "Superseded narration"
        case .currentTakes: "Current narration"
        case .assembledWithRecipe: "Assembled tapes"
        case .assembledWithoutRecipe: "Assembled tapes without a recipe"
        case .assembledRetiredVoice: "Assembled tapes in a retired voice"
        case .recycleBin: "Recently deleted"
        case .voicePreviews: "Voice previews"
        }
    }

    /// What deleting this actually costs. Never "nothing" unless it is nothing.
    public var consequence: String {
        switch self {
        case .supersededTakes:
            "Already out of date. The queue re-renders these anyway, so this costs nothing."
        case .currentTakes:
            "Still valid. Deleting them means rendering them again before the next assembly."
        case .assembledWithRecipe:
            "Rebuildable exactly from the recipe that made them."
        case .assembledWithoutRecipe:
            "These predate frozen recipes. They can be built again from their plan, but not identically."
        case .assembledRetiredVoice:
            "Spoken by a voice that is no longer installed. They still play, but nothing can make them again — rebuilding produces a different voice."
        case .recycleBin:
            "Already deleted, kept for thirty days. Emptying now is permanent."
        case .voicePreviews:
            "A minute to render again."
        }
    }

    /// True when nothing at all is lost -- not merely nothing important.
    public var costsNothing: Bool { self == .supersededTakes }
}

public struct StorageReport: Sendable {
    public struct Group: Sendable {
        public var kind: StorageKind
        public var files: [URL]
        public var bytes: Int64

        /// Public so a report can be *constructed* rather than only measured.
        /// The cross-platform fixture builds one over a scratch tree, because
        /// `purge` deletes and cannot be exercised against the real library.
        public init(kind: StorageKind, files: [URL], bytes: Int64) {
            self.kind = kind; self.files = files; self.bytes = bytes
        }
        public var count: Int { files.count }
    }

    public var groups: [Group]
    /// Everything the application owns, including what it will never delete.
    public var totalBytes: Int64

    public init(groups: [Group], totalBytes: Int64) {
        self.groups = groups; self.totalBytes = totalBytes
    }

    public func group(_ kind: StorageKind) -> Group? {
        groups.first { $0.kind == kind }
    }
    public var reclaimableBytes: Int64 { groups.reduce(0) { $0 + $1.bytes } }

    public static func format(_ bytes: Int64) -> String {
        let f = ByteCountFormatter()
        f.allowedUnits = [.useMB, .useGB]
        f.countStyle = .file
        return f.string(fromByteCount: bytes)
    }
}

public enum StorageAudit {
    private static func size(of url: URL) -> Int64 {
        (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize).flatMap(Int64.init) ?? 0
    }

    /// Sources the render queue can rebuild a take from, keyed by output name.
    private static func sourcesByOutput(root: URL, library: Library) -> [String: String] {
        var out: [String: String] = [:]
        // Every authored file, at every verbosity it was written at: a
        // segment with a v1 and a v3 body produces two differently named
        // takes, and missing one would report a live take as orphaned.
        var files: [URL] = library.focus.flatMap(\.scripts)
        for ref in library.segments + library.continuousSegments {
            files.append(ref.url)
            files.append(contentsOf: ref.verbosityFiles.values)
        }
        for file in files {
            guard let source = try? String(contentsOf: file, encoding: .utf8) else { continue }
            for item in RenderPlan.items(gwsFile: file, source: source) {
                out[item.outputName] = source
            }
        }
        return out
    }

    /// Measure. Reads the disk and decides nothing on the listener's behalf.
    public static func measure(root: URL, library: Library, renderKey: String,
                               voice: String,
                               fileManager fm: FileManager = .default) -> StorageReport {
        var groups: [StorageKind: (files: [URL], bytes: Int64)] = [:]
        func add(_ kind: StorageKind, _ url: URL) {
            groups[kind, default: ([], 0)].files.append(url)
            groups[kind, default: ([], 0)].bytes += size(of: url)
        }

        // Narration takes, split by whether they are still worth anything.
        //
        // **Every voice's directory, not just the selected one.** Takes live
        // under `segments-rendered/<voice>/`, so retiring a voice leaves its
        // renders behind -- a gigabyte of them, in the case that prompted
        // this -- and measuring only the current voice made exactly the pile
        // this panel exists to find invisible. A take belonging to a voice
        // that no longer exists is superseded by definition: nothing can use
        // it and nothing will re-render it.
        let sources = sourcesByOutput(root: root, library: library)
        let liveRenders = Set(library.focus.flatMap(\.renders).map(\.lastPathComponent))
        let known = Set(library.voices.map(\.name))
        let renderedRoot = root.appending(path: "segments-rendered")
        let voiceDirs = ((try? fm.contentsOfDirectory(at: renderedRoot,
                                                      includingPropertiesForKeys: nil)) ?? [])
            .filter(\.hasDirectoryPath)
        for takeDir in voiceDirs {
        let owner = takeDir.lastPathComponent
        let retired = !known.contains(owner)
        for url in (try? fm.contentsOfDirectory(at: takeDir, includingPropertiesForKeys: nil)) ?? []
        where url.pathExtension == "wav" {
            if retired { add(.supersededTakes, url); continue }
            let name = url.lastPathComponent
            if let source = sources[name] {
                let current = owner == voice
                    && RenderPlan.isCurrent(name, source: source,
                                            in: takeDir, renderKey: renderKey)
                add(current ? .currentTakes : .supersededTakes, url)
            } else if let session = announcementSession(name), liveRenders.contains(session) {
                // An announcement belongs to one assembled tape and is current
                // for as long as that tape is.
                add(.currentTakes, url)
            } else {
                // Nothing in the library produces this name any more.
                add(.supersededTakes, url)
            }
        }
        }

        // Assembled tapes. The audio only -- never the folder, never the notes.
        let installedVoices = Set(library.voices.map(\.name))
        let recipes = Set(((try? fm.contentsOfDirectory(
            at: root.appending(path: "memory/sessions"), includingPropertiesForKeys: nil)) ?? [])
            .filter { $0.pathExtension == "json" }
            .map { $0.deletingPathExtension().lastPathComponent })
        for dir in library.focus.flatMap(\.renders) {
            let wav = dir.appending(path: "session.wav")
            guard fm.fileExists(atPath: wav.path) else { continue }
            // **Which voice spoke it outranks whether a recipe exists.**
            // A recipe promises an exact rebuild, and that promise is void the
            // moment the voice it names is gone: the words come back, the
            // voice does not. Retiring the first snepssen made all thirty
            // tapes on this disk unreproducible at once, while the panel went
            // on offering eleven of them as "rebuildable exactly".
            let spokenBy = SessionManifestIO.load(dir.appending(path: "manifest.json"))?.voice
            if let spokenBy, !spokenBy.isEmpty, !installedVoices.contains(spokenBy) {
                add(.assembledRetiredVoice, wav)
            } else {
                add(recipes.contains(dir.lastPathComponent)
                    ? .assembledWithRecipe : .assembledWithoutRecipe, wav)
            }
        }

        // The thirty-day bin, and previews.
        if let walker = fm.enumerator(at: root.appending(path: "memory/deleted"),
                                      includingPropertiesForKeys: nil) {
            for case let url as URL in walker where url.lastPathComponent != "index.json" {
                // Same reasoning as `purge`: the filesystem decides, not the
                // URL's spelling. An enumerator usually sets the hint, but
                // "usually" is not a contract and the cost of being wrong here
                // is a directory in a delete list.
                var isDirectory: ObjCBool = false
                guard fm.fileExists(atPath: url.path, isDirectory: &isDirectory),
                      !isDirectory.boolValue else { continue }
                add(.recycleBin, url)
            }
        }
        for voiceRef in library.voices {
            let preview = voiceRef.dir.appending(path: "preview.wav")
            if fm.fileExists(atPath: preview.path) { add(.voicePreviews, preview) }
        }

        let ordered = StorageKind.allCases.compactMap { kind -> StorageReport.Group? in
            guard let g = groups[kind], !g.files.isEmpty else { return nil }
            return StorageReport.Group(kind: kind, files: g.files, bytes: g.bytes)
        }
        return StorageReport(groups: ordered, totalBytes: directorySize(root, fileManager: fm))
    }

    /// `<session-id>-announcement.takeN.wav` -> `<session-id>`.
    /// Exposed for the cross-platform fixture: the naming rule is shared with
    /// the TypeScript port and has to be pinned somewhere both can see it.
    public static func announcementSessionForTests(_ name: String) -> String? {
        announcementSession(name)
    }

    private static func announcementSession(_ name: String) -> String? {
        guard let range = name.range(of: "-announcement.take") else { return nil }
        return String(name[name.startIndex..<range.lowerBound])
    }

    public static func directorySize(_ url: URL, fileManager fm: FileManager = .default) -> Int64 {
        var total: Int64 = 0
        if let walker = fm.enumerator(at: url, includingPropertiesForKeys: [.fileSizeKey]) {
            for case let file as URL in walker where !file.hasDirectoryPath {
                total += size(of: file)
            }
        }
        return total
    }

    /// Delete exactly what the report named for these kinds, and nothing else.
    ///
    /// Takes the report rather than re-deriving the list, so that what the
    /// listener was shown and what is removed cannot come apart between one
    /// and the other.
    @discardableResult
    public static func purge(_ report: StorageReport, kinds: Set<StorageKind>,
                             fileManager fm: FileManager = .default) -> Int64 {
        var freed: Int64 = 0
        for group in report.groups where kinds.contains(group.kind) {
            for url in group.files {
                // A directory here would mean the report had strayed; refuse
                // rather than recurse.
                //
                // **Ask the filesystem, not the URL.** `hasDirectoryPath`
                // reports how a URL was *spelled* -- whether it was built with
                // a trailing slash or a directory hint -- not what is on disk.
                // A real directory named the way a report names a file returns
                // false, so this guard let it through and `removeItem` removed
                // it recursively, taking the `notes.md` inside with it. That is
                // exactly the one thing this file promises never to do.
                //
                // Found by the cross-platform port: a scratch tree with a
                // directory deliberately listed among the files, which no
                // measured report will ever produce and so nothing here had
                // ever exercised.
                var isDirectory: ObjCBool = false
                let exists = fm.fileExists(atPath: url.path, isDirectory: &isDirectory)
                guard exists, !isDirectory.boolValue else { continue }
                let bytes = size(of: url)
                if (try? fm.removeItem(at: url)) != nil { freed += bytes }
            }
        }
        return freed
    }
}
